src/Widget.php

```diff
@@ -0,0 +1,192 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Widget;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Yiisoft\Widget\Exception\InvalidConfigException;
+use Yiisoft\Widget\Event\AfterRun;
+use Yiisoft\Widget\Event\BeforeRun;
+
+/**
+ * Widget generates a string content based on some logic and input data.
+ * These are typically used in templates to conceal complex HTML rendering logic.
+ *
+ * This is the base class that is meant to be inherited when implementing your own widgets.
+ */
+abstract class Widget
+{
+    protected EventDispatcherInterface $eventDispatcher;
```

Comment:
```suggestion
    protected ?EventDispatcherInterface $eventDispatcher = null;
```

Reply:
Widget is configurable since you can extend it and simply if you have more dependencies in the constructor of your widget you pass it to the base constructor `parent::__construct($eventDispatcher)`, on the other hand we use the hiqdev/composer-config-plugin that the widget will already come with the factory configuration in tests and common, and you can use the yiisoft/factory with any container without being tied to the framework, particularly I think it is better to inject the dependencies into the constructor.